samples/net/coap_server/src/coap-server.c

```diff
@@ -0,0 +1,1310 @@
+			return -EINVAL;
+		}
+
+		coap_pending_cycle(pending);
+		pending = coap_pending_next_to_expire(pendings, NUM_PENDINGS);
+
+		k_delayed_work_submit(&retransmit_work, pending->timeout);
+	}
+
+	return net_context_sendto(pkt, (const struct sockaddr *)&from,
+				  sizeof(struct sockaddr_in6),
+				  NULL, 0, NULL, NULL);
+}
+
+static int large_get(struct coap_resource *resource,
+		     struct coap_packet *request)
+{
+	static struct coap_block_context ctx;
+	struct net_pkt *pkt;
+	struct net_buf *frag;
+	struct sockaddr_in6 from;
+	struct coap_packet response;
+	u8_t token[8], code, type;
+	u8_t payload[64];
+	u16_t size;
+	u16_t id;
+	u8_t tkl;
+	int r;
+
+	if (ctx.total_size == 0) {
+		coap_block_transfer_init(&ctx, COAP_BLOCK_64,
+					 BLOCK_WISE_TRANSFER_SIZE_GET);
+	}
+
+	r = coap_update_from_block(request, &ctx);
+	if (r < 0) {
+		return -EINVAL;
+	}
+
+	get_from_ip_addr(request, &from);
+	code = coap_header_get_code(request);
+	type = coap_header_get_type(request);
+	id = coap_header_get_id(request);
+	tkl = coap_header_get_token(request, token);
+
+	NET_INFO("*******\n");
+	NET_INFO("type: %u code %u id %u\n", type, code, id);
+	NET_INFO("*******\n");
+
+	pkt = net_pkt_get_tx(context, K_FOREVER);
+	frag = net_pkt_get_data(context, K_FOREVER);
+
+	net_pkt_frag_add(pkt, frag);
+
+	r = coap_packet_init(&response, pkt, 1, COAP_TYPE_ACK,
+			     tkl, (u8_t *) token,
+			     COAP_RESPONSE_CODE_CONTENT, id);
+	if (r < 0) {
+		return -EINVAL;
+	}
+
+	r = coap_packet_append_option(&response, COAP_OPTION_CONTENT_FORMAT,
+				      &plain_text_format,
+				      sizeof(plain_text_format));
+	if (r < 0) {
+		return -EINVAL;
+	}
+
+	r = coap_append_block2_option(&response, &ctx);
+	if (r < 0) {
+		return -EINVAL;
+	}
+
+	r = coap_packet_append_payload_marker(&response);
+	if (r) {
+		net_pkt_unref(pkt);
+		return -EINVAL;
+	}
+
+	size = min(coap_block_size_to_bytes(ctx.block_size),
+		   ctx.total_size - ctx.current);
+
+	memset(payload, 'A', size);
+
+	r = coap_packet_append_payload(&response, (u8_t *)payload, size);
+	if (r < 0) {
+		net_pkt_unref(pkt);
+		return -EINVAL;
+	}
+
+	r = coap_next_block(&response, &ctx);
+	if (!r) {
+		/* Will return 0 when it's the last block. */
+		memset(&ctx, 0, sizeof(ctx));
+	}
+
+	return net_context_sendto(pkt, (const struct sockaddr *)&from,
+				  sizeof(struct sockaddr_in6),
+				  NULL, 0, NULL, NULL);
+}
+
+static int large_update_put(struct coap_resource *resource,
+			    struct coap_packet *request)
+{
+	static struct coap_block_context ctx;
+	struct net_pkt *pkt;
+	struct net_buf *frag;
+	struct sockaddr_in6 from;
+	struct coap_packet response;
+	u8_t token[8];
+	u16_t offset;
+	u16_t id;
+	u8_t payload;
+	u8_t code;
+	u8_t type;
+	u8_t tkl;
+	int r;
+
+	frag = net_frag_skip(request->frag, request->offset, &offset,
+			     request->hdr_len + request->opt_len);
+	if (!frag && offset == 0xffff) {
+		return -EINVAL;
+	}
+
+	frag = net_frag_read_u8(frag, offset, &offset, &payload);
```

Comment:
It seems that you changed the semantics of the large_update_put().
It was returning errors when there's no payload.
But here (line 726~730) changes it to return an error when there is payload.

Originally,
```
        payload = zoap_packet_get_payload(request, &len);
        if (!payload) {
                NET_ERR("Packet without payload\n");
                return -EINVAL;
        }
```

Also, would it be nice if CoAP library could provide helper functions to get the payload offset and payload length?

Reply:
According to samples/net/coap_server/README.rst
--------------------
Use this command on the host to run the`libcoap`_ implementation of
the ETSI test cases:

.. code-block:: console

   sudo ./examples/etsi_coaptest.sh -i tap0 2001:db8::1
----------------------------------------------------------------------

Tests were failed because of original functionality. So I introduced these changes and it continues. Either these sample is not tested against latest libcoap or something else.
That's why I added those 'TODO' comments.

Reply:
Yup, you are right. The test-case is not passed at the original zoap_server implementation.
It returns nothing when payload doesn't exist which is just the case when you run the test case.

Command `./coap-client -m put -b 1024 coap://[2001:db8::1]/large-update`
issues an PUT request w/ block1 options (size=1024) and zero payload.

I guess the correct handling for this should be checking the "MORE" flag of the block1 option.
If there's no more block, then empty payload should be allowed. Otherwise, it should be treated as an error.

Also, I think you should put these changes to the description of the commit.
It's hard to tell whether it's a simply porting or a porting w/ bug fixing.